logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java

```diff
@@ -0,0 +1,211 @@
+ * until they are handed off to the consumer.
+ */
+public final class Sequencer implements LogStreamBatchWriter, Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(Sequencer.class);
+  private final int partitionId;
+  private final int maxFragmentSize;
+
+  private volatile long position;
+  private volatile boolean isClosed = false;
+  private volatile ActorCondition consumer;
+  private final Queue<SequencedBatch> queue = new ArrayBlockingQueue<>(128);
+  private final ReentrantLock lock = new ReentrantLock();
+  private final SequencerMetrics metrics;
+
+  public Sequencer(final int partitionId, final long initialPosition, final int maxFragmentSize) {
+    LOG.trace("Starting new sequencer at position {}", initialPosition);
+    this.position = initialPosition;
+    this.partitionId = partitionId;
+    this.maxFragmentSize = maxFragmentSize;
+    this.metrics = new SequencerMetrics(partitionId);
+  }
+
+  /**
+   * @param eventCount the potential event count we want to check
+   * @param batchSize the potential batch Size (in bytes) we want to check
+   * @return True if the serialized batch would fit within {@link Sequencer#maxFragmentSize}, false
+   *     otherwise.
+   */
+  @Override
+  public boolean canWriteEvents(final int eventCount, final int batchSize) {
+    final int framedMessageLength =
+        batchSize
+            + eventCount * (DataFrameDescriptor.HEADER_LENGTH + FRAME_ALIGNMENT)
+            + FRAME_ALIGNMENT;
+    return framedMessageLength <= maxFragmentSize;
+  }
+
+  /**
+   * @param appendEntry the entry to write
+   * @param sourcePosition a back-pointer to the record whose processing created this entry
+   * @return -1 if write was rejected, the position of the entry if write was successful.
+   */
+  @Override
+  public long tryWrite(final LogAppendEntry appendEntry, final long sourcePosition) {
+    if (isClosed) {
+      LOG.warn("Rejecting write of {}, sequencer is closed", appendEntry);
+      return -1;
+    }
+
+    if (!isEntryValid(appendEntry)) {
+      LOG.warn("Reject write of invalid entry {}", appendEntry);
+      return 0;
+    }
+
+    final boolean isEnqueued;
+    final long currentPosition;
+
+    lock.lock();
+    try {
+      currentPosition = position;
+      isEnqueued =
+          queue.offer(new SequencedBatch(currentPosition, sourcePosition, List.of(appendEntry)));
+      if (isEnqueued) {
+        position = currentPosition + 1;
+      }
+    } finally {
+      lock.unlock();
+    }
+
+    if (consumer != null) {
+      consumer.signal();
+    }
+    metrics.observeBatchSize(1);
+    metrics.setQueueSize(queue.size());
+    if (isEnqueued) {
+      return currentPosition;
+    } else {
+      LOG.trace("Rejecting write of {}, sequencer queue is full", appendEntry);
+      return -1;
+    }
+  }
+
+  /**
+   * @param appendEntries a set of entries to append; these will be appended in the order in which
```

Comment:
:upside_down_face: I don't notice changes between this and the interface docs, so we can just use the `{@inheritDoc}` tag, I think.